### PR TITLE
Allow for generated files

### PR DESCRIPTION
**Pull request by dmadisetti**
Sorry I'm back

Allows for generated files to be checked by mypy, and for external generated files to be ignored.

I've had this patch since my last PR, but only just got round to writing some tests. This might sound super edge-case and useless, but arose from me implementing this project with protobuf

If for some reason, a generated python file `is` meant to be explicitly checked, then we handle that case too.

### CODE REVIEW

**Comment by thundergolfer** [comment]
> Sorry I'm back

Haha don't apologise! It's awesome to have contributors.

File: templates/mypy.sh.tpl [hunk 1]
```diff
@@ -27,7 +27,7 @@ main() {
   fi
 
   set +o errexit
-  output=$($mypy {VERBOSE_OPT} --bazel --package-root . --config-file {MYPY_INI_PATH} --cache-map {CACHE_MAP_TRIPLES} -- {SRCS} 2>&1)
+  output=$($mypy {VERBOSE_OPT} --bazel --package-root . --package-root bazel-out/k8-fastbuild/bin --config-file {MYPY_INI_PATH} --cache-map {CACHE_MAP_TRIPLES} -- {SRCS} 2>&1)
```
Comment (thundergolfer) [inline templates/mypy.sh.tpl]: `bazel-out/k8-fastbuild/bin` I don't think that can be hardcoded.

If I do an `ls` on `bazel-out/` I see: 

```
_tmp  darwin-fastbuild	host  stable-status.txt  volatile-status.txt
```

**Comment by thundergolfer** [comment]
> This might sound super edge-case and useless, but arose from me implementing this project with protobuf

Can you expand on this? You were trying to implement this integration's functionality with protobuf?  I don't quite understand.

File: templates/mypy.sh.tpl (hunk 1 above)
Comment (dmadisetti) [reply templates/mypy.sh.tpl]: Good catch, I didn't know that. I'll look into cross platform methods

**Comment by dmadisetti** [comment]
> > This might sound super edge-case and useless, but arose from me implementing this project with protobuf
> 
> Can you expand on this? You were trying to implement this integration's functionality with protobuf? I don't quite understand.

Yeah! So I'm just trying to type-check some code that implements [protobuf](https://github.com/bazelbuild/rules_proto), which generates python for data serialization. My BUILD looks something like:

```starlark
load(
    "@build_stack_rules_proto//python:python_proto_library.bzl",
    "py_proto_library",
)
py_proto_library(
    name = "py_proto",
    deps = ["//configs:proto"],
)
py_test(
    name = "pd3_test",
    srcs = ["pd3_test.py"],
    deps = [
        ":py_proto",
        "//pd3",
    ],
) 

```
Without the added changes, the invoked mypy command is:
```bash
external/mypy_integration/mypy/mypy --verbose --bazel --package-root .  --config-file external/mypy_integration_config/mypy.ini --cache-map bazel-out/k8-fastbuild/bin/configs/config_pb2.py bazel-out/k8-fastbuild/bin/configs/config_pb2.py.meta.json bazel-out/k8-fastbuild/bin/configs/config_pb2.py    .data.json python/pd3.py python/pd3.py.meta.json python/pd3.py.data.json python/utils.py python/utils.py.meta.json python/utils.py.data.json python/pd3_test.py python/pd3    _test.py.meta.json python/pd3_test.py.data.json -- bazel-out/k8-fastbuild/bin/configs/config_pb2.py python/pd3.py python/utils.py python/pd3_test.py
```

and breaks with `mypy: can't read file 'bazel-out/k8-fastbuild/bin/configs/config_pb2.py': No such file or directory`. I think it's fair that generated file from a dependency should not be typed (same motivation for excluding external sources), so we can just remove it from the command. A clear indication that this file was generated is the fact the path is under `bazel-out/k8-fastbuild/bin/` (or darwin-fastbuild for macs I guess, who knows for windows) not relative to the files, so we just filter for that.

However, if for some reason the generated file is included in srcs, it should be typed.

There may be a better way, but this seems reasonable-ish? I could play around with the examples folder if I get time

**Comment by dmadisetti** [comment]
:thinking: This is starting to look more like a bit of a hack. Let me know if you have other ideas. Fairly certain that bazel uses "/" on windows as well.

**Comment by thundergolfer** [comment]
> However, if for some reason the generated file is included in srcs, it should be typed.

What could that reason be? If it's generated code, it should never be included in the `srcs` attr of a `py_*` rule right?

**Comment by dmadisetti** [comment]
The tests I provided do exactly this: https://github.com/thundergolfer/bazel-mypy-integration/pull/11/files#diff-c40927b6657800717d4b0b498e1bad55R53

__Can__ vs **should** is an important distinction though. I'd agree it's not good practice, but mypy shouldn't shouldn't break inexplicably.

**Comment by dmadisetti** [comment]
Additional changes are for compat with https://github.com/thundergolfer/bazel-python-mypy-protobuf/pull/2

Tests pass and the project builds

**Comment by ashwin153** [comment]
@thundergolfer Any reason why this can't be merged? The integration is currently unusable in repositories that use Protocol Buffers / gRPC.

**Comment by dmadisetti** [comment]
If I resolve do you think we can get this in? About 3 years now, and I just ran into this problem again.
